# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.cpp
# Commit: 67ae30727ed8
# Full Hash: 67ae30727ed835cf73153a4ee8f0c9c7931aab09
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2021-12-23 09:34:36
# Regressor Bug: 1642412
# File Overlap Count: 7
# Description:
#   Bug 1642412 - wasm: Remove unnecessary code from Instance for inline table methods. r=lth
#   
#   This commit removes support for anyref in Instance::tableGet/tableSet,
#   making it clear that these functions are only used for function tables.
#   Instance::tableSize is removed as no longer needed.
# ==============================================================================

diff -r 90fa382b5a20 -r 67ae30727ed8 js/src/wasm/WasmInstance.cpp
--- a/js/src/wasm/WasmInstance.cpp	Wed Dec 22 17:36:30 2021 +0000
+++ b/js/src/wasm/WasmInstance.cpp	Wed Dec 22 17:36:30 2021 +0000
@@ -1175,9 +1175,9 @@
   return 0;
 }
 
-/* static */ void* Instance::tableGet(Instance* instance, uint32_t index,
-                                      uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableGet.failureMode == FailureMode::FailOnInvalidRef);
+/* static */ void* Instance::tableGetFunc(Instance* instance, uint32_t index,
+                                          uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableGetFunc.failureMode == FailureMode::FailOnInvalidRef);
 
   JSContext* cx = instance->tlsData()->cx;
   const Table& table = *instance->tables()[tableIndex];
@@ -1188,20 +1188,13 @@
     return AnyRef::invalid().forCompiledCode();
   }
 
-  switch (table.repr()) {
-    case TableRepr::Ref:
-      return table.getAnyRef(index).forCompiledCode();
-    case TableRepr::Func: {
-      MOZ_RELEASE_ASSERT(!table.isAsmJS());
-      RootedFunction fun(cx);
-      if (!table.getFuncRef(cx, index, &fun)) {
-        return AnyRef::invalid().forCompiledCode();
-      }
-      return FuncRef::fromJSFunction(fun).forCompiledCode();
-    }
+  MOZ_RELEASE_ASSERT(table.repr() == TableRepr::Func);
+  MOZ_RELEASE_ASSERT(!table.isAsmJS());
+  RootedFunction fun(cx);
+  if (!table.getFuncRef(cx, index, &fun)) {
+    return AnyRef::invalid().forCompiledCode();
   }
-
-  MOZ_CRASH("Should not happen");
+  return FuncRef::fromJSFunction(fun).forCompiledCode();
 }
 
 /* static */ uint32_t Instance::tableGrow(Instance* instance, void* initValue,
@@ -1248,9 +1241,9 @@
   MOZ_CRASH("Should not happen");
 }
 
-/* static */ int32_t Instance::tableSet(Instance* instance, uint32_t index,
-                                        void* value, uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableSet.failureMode == FailureMode::FailOnNegI32);
+/* static */ int32_t Instance::tableSetFunc(Instance* instance, uint32_t index,
+                                            void* value, uint32_t tableIndex) {
+  MOZ_ASSERT(SASigTableSetFunc.failureMode == FailureMode::FailOnNegI32);
 
   JSContext* cx = instance->tlsData()->cx;
   Table& table = *instance->tables()[tableIndex];
@@ -1261,30 +1254,17 @@
     return -1;
   }
 
-  switch (table.repr()) {
-    case TableRepr::Ref:
-      table.fillAnyRef(index, 1, AnyRef::fromCompiledCode(value));
-      break;
-    case TableRepr::Func:
-      MOZ_RELEASE_ASSERT(!table.isAsmJS());
-      if (!table.fillFuncRef(Nothing(), index, 1,
-                             FuncRef::fromCompiledCode(value), cx)) {
-        ReportOutOfMemory(cx);
-        return -1;
-      }
-      break;
+  MOZ_RELEASE_ASSERT(table.repr() == TableRepr::Func);
+  MOZ_RELEASE_ASSERT(!table.isAsmJS());
+  if (!table.fillFuncRef(Nothing(), index, 1,
+                         FuncRef::fromCompiledCode(value), cx)) {
+    ReportOutOfMemory(cx);
+    return -1;
   }
 
   return 0;
 }
 
-/* static */ uint32_t Instance::tableSize(Instance* instance,
-                                          uint32_t tableIndex) {
-  MOZ_ASSERT(SASigTableSize.failureMode == FailureMode::Infallible);
-  Table& table = *instance->tables()[tableIndex];
-  return table.length();
-}
-
 /* static */ void* Instance::refFunc(Instance* instance, uint32_t funcIndex) {
   MOZ_ASSERT(SASigRefFunc.failureMode == FailureMode::FailOnInvalidRef);
   JSContext* cx = instance->tlsData()->cx;